File: 파이썬 알고리즘 인터뷰/7장 선형 자료구조/세 수의 합.js
// 나의 투포인터를 이용한 풀이(Runtime: 468ms, Memory: 52.3MB)
const threeSum = function(nums) {
  const sorted = nums.sort((numA, numB) => numA - numB);
  const answer = {};
  
  for (let idx = 0; idx < nums.length; idx++) {
    let left = idx + 1;
    let right = nums.length - 1;
    
    // [참고] outOfRange는 아래 조건만으로 예외처리된다.
    while(left < right) {
      const result = nums[idx] + nums[left] + nums[right];
      
      if (result < 0) {
        left++;
      } else if (result > 0) {
        right--;
      } else {
        // 중복 처리를 위해 숫자를 문자열로 변환해 더한 값을 Key로 사용한다.
        const tripet = [nums[idx], nums[left], nums[right]]; 
        const key = tripet.reduce((str, num) => str + num, "");
        
        answer[key] = tripet;
        left++;
        right--;
      }
    }
  }
   
  return Object.values(answer);
};

File: 파이썬 알고리즘 인터뷰/7장 선형 자료구조/세 수의 합.py
class Solution:
    # 나의 브루트포스를 이용한 풀이 (Time Limit Exceeded)
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        answer = []
        
        for i in range(0, len(nums)):
            for j in range(i + 1, len(nums)):
                for k in range(j + 1, len(nums)):
                    if nums[i] + nums[j] + nums[k] == 0:
                        sorted_array = sorted([nums[i], nums[j], nums[k]])
                        
                        if sorted_array not in answer:
                            answer.append(sorted_array)
        
        return answer

    # 나의 투포인터를 이용한 풀이(Runtime: 4732ms, Memory: 17.5MB)
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        answer = []
        
        # 배열을 정렬해 투 포인터를 사용할 수 있도록 한다.
        nums.sort()
        
        for idx in range(0, len(nums)):
          left = idx + 1
          right = len(nums) - 1
          
          # 투 포인터가 만날 때까지 반복한다.
          while left < right:
              temp = nums[idx] + nums[left] + nums[right]
              
              # 합이 0보다 작은 경우, 왼쪽 포인터만 이동시킨다.
              if temp < 0:
                  left += 1
              # 합이 0보다 큰 경우, 오른쪽 포인터만 이동시킨다.
              elif temp > 0:
                  right -= 1
              # 합이 0인 경우, answer 배열에 추가하고 양측 포인터를 모두 이동시킨다.
              else:
                  if [nums[idx], nums[left], nums[right]] not in answer:
                      answer.append([nums[idx], nums[left], nums[right]])
                  left += 1
                  right -= 1
                      
        return answer